README.rst

What is this project?
^^^^^^^^^^^^^^^^^^^^^

LyfeShoppe

Installation instructions
^^^^^^^^^^^^^^^^^^^^^^^^^

This process will be more streamlined in the future but for now you're on your
own for setting up virtualenv, installing Docker, installing nodejs and doing
the post-install steps.

Install Virtualenv
''''''''''''''''''

- https://virtualenv.pypa.io/en/latest/installation.html

Install Docker
''''''''''''''

- https://docs.docker.com/installation (docker itself)
- https://github.com/docker/compose/releases (docker-compose tool for development)

Make sure you get ``docker-compose v1.4.0+``.

Install nodejs
''''''''''''''

- https://nodejs.org/download/ (runtime dependency for assets)

Register on Stripe to get your API keys
'''''''''''''''''''''''''''''''''''''''

- https://dashboard.stripe.com/register
- Go to your account settings on Stripe
- Copy your API keys
- Follow the directions in ``config/settings.py``

Review the plans in the above settings file before proceeding. You can always
change or delete plans later, so don't sweat it. Just make sure you know
what's being created.

I installed everything, now what?
'''''''''''''''''''''''''''''''''

Create a bit of supporting directory structure
----------------------------------------------

- Create a new folder somewhere, let's say: ``/tmp/lyfeshoppe``
- Create a new ``website`` folder inside of that: ``/tmp/lyfeshoppe/website``
- CD into that directory: ``cd /tmp/lyfeshoppe/website``

Clone the repo and install all dependencies
-------------------------------------------

- Clone this repo ``git clone git@github.com:himudianda/lyfeshoppe.git``
- Type ``cd lyfeshoppe``
- Activate your virtualenv
- Type ``pip install --editable .`` to activate the CLI
- Type ``pip install -r requirements.txt`` to install dependencies
- Type ``npm install`` to install the asset dependencies

Set up docker-compose
---------------------

- Edit ``docker-compose.yml`` and `setup a local volume for Postgres/Redis`__
- Type ``docker-compose up`` to start Postgres/Redis

Initialize everything and view the app
--------------------------------------

- Open a new terminal window
- Type ``run assets build`` to create the build directory and manifest file
- Type ``run`` to see a list of what's available
- Type ``run db reset lyfeshoppe lyfeshoppe_test`` to initialize the databases
- Type ``run stripe sync_plans`` to sync your ``STRIPE_PLANS`` to the gateway
- Type ``run all`` to start everything
- Visit http://localhost:8000 in your browser
- If you wish to login, email: ``dev@localhost.com`` / password: ``password``

Do I need to do the above steps all the time?
'''''''''''''''''''''''''''''''''''''''''''''

Heck no, from now on you only need to do this:

- Activate your virtualenv
- Type ``run all`` to start everything
- Visit http://localhost:8000 in your browser

Also in the future I will be working on ways to automate the above steps. You
have to remember the above steps are replacing things like Vagrant and more.

You just setup a complete dev environment which you can reproduce on
any machine capable of running Docker. Your environment is now very close to
what we will be running in production.

How do I shut everything down?
''''''''''''''''''''''''''''''

- Hit CTRL+C a few times to stop everything
- Type ``docker-compose stop`` to ensure all containers are stopped
- Confirm no containers are running by typing ``docker ps``

How do I provide my own settings?
'''''''''''''''''''''''''''''''''

- Create a ``settings.py`` file at ``$PROJECT_ROOT/instance/settings.py``
- Overwrite as many settings as you want
View the `example production settings.py <https://github.com/himudianda/lyfeshoppe/blob/master/instance/settings.py.production_example>`_.

Can I quickly change my schema without migrating?
'''''''''''''''''''''''''''''''''''''''''''''''''

Yep, just be warned that this will completely purge your database but doing
this early on in development can sometimes be reasonable while you tinker with
your schema very frequently.

- Shut everything down
- Type ``docker-compose run postgres``
- Type ``run db reset lyfeshoppe lyfeshoppe_test``
- Type ``run add all``

This will drop your database, create a new one and seed it with fake data.

My billing history is always empty
''''''''''''''''''''''''''''''''''

Filling out the billing history requires setting up webhooks with Stripe. You
can do that in your Stripe account dashboard under webhooks.

You will need to setup something like ngrok so localhost is accessible outside
of your local network. It does this by setting up a tunnel.

Also make sure to look at the comments in ``config/settings.py`` for the
``SERVER_NAME`` setting.

How can I test the Twitter stream?
''''''''''''''''''''''''''''''''''

After everything is running and your settings are configured just type ``run stream broadcast``
and it will start reading in events from Twitter and broadcast the messages to the
websocket server.

You can bypass broadcasting and simply listen it on the stream by typing
``run stream listen`` instead.

Learn more
^^^^^^^^^^

What packages are being used?
'''''''''''''''''''''''''''''

Check the commented ``requirements.txt`` for package specifics.

How will the project be managed?
''''''''''''''''''''''''''''''''

Upcoming features
-----------------

I'm an organized person but not OCD about it. I will do my best to add pending
features to the issue tracker with a specific label. Not all features will get
added to the issue tracker because who wants to write issues all day!

Branches
--------

The latest "unstable but might be stable" version will be master.

Stable releases will be tagged and released using the http://semver.org/ system.
However early on in the project there's a very good chance the versions will
not adhere to semver perfectly.

About the author
^^^^^^^^^^^^^^^^

- Harshit Imudianda | `Website <http://imudianda.com>`_ | `LinkedIn <https://www.linkedin.com/in/harshitimudianda>`_ | `GitHub <https://github.com/himudianda>`_


CHANGES.rst
Version 0.5 (2015-11-29)
^^^^^^^^^^^^^^^^^^^^^^^^^^
- Registrations/Logins only happen on User models (i.e. no register/login of Business models)
- Got rid of Product models & simplified the Business models - also removed some views related to these models.
- All Admin pages styled to dashboard theme


Version 0.4 (2015-11-27)
^^^^^^^^^^^^^^^^^^^^^^^^^^
- Admin dashboard now has ability to view all Businesses & Products
- You can also bulk delete, edit & add new products/businesses from the admin view


Version 0.3 (2015-11-26)
^^^^^^^^^^^^^^^^^^^^^^^^^^
- Modified all the to-be-used views of LyfeShoppe to either dashboard or frontend theme layout.
- Extensive frontend work - login, signup, home, FAQs, Support etc pages are all modified with new layout.
- Surfing through LyfeShoppe should give a nice monolithic experience.


Version 0.2 (2015-11-21)
^^^^^^^^^^^^^^^^^^^^^^^^^^
- Added 2 new themes for frontend & dashboard
- Added sample templates & views for the dashboard & frontend theme demos.
- These themes were transformed using lyfeshoppe_private/scripts/transform_themes.py version 0.1


Version 0.1 (2015-11-14)
^^^^^^^^^^^^^^^^^^^^^^^^^^
- Major features & infrastructure work has been implemented as follows.

Features:
- Implemented User roles, subscriptions, Issue tracking, Coupons, Credit Card Handling, etc.
- Stripe API used for billing
- All subscribed users can receive tweets about cats :)
- Unit Tests have been implemented for testing core features.
- Issue tracking & support has also been added & can be viewed on Admin interface
- Tools like flower, cAdvisor used for better system tracking.

Packages:
- Flask-Mail used to send out emails
- Redis is used for multiple things: pub-sub part in websocket server, to back celery & to back flask-cache as the backend.
- Celery is the background task queue & Flower is the front end to view celery tasks
- Flask-Cache helps caching an entire flask view OR a jinja2 snippet.
- flask-bouncer is used for handling permissions - i.e. showing diff data for different User roles
- Subscription service implemented using Stripe APIs
- Tweepy is used for getting realtime tweets
- Unit Tests implemented using pytest, pytest-cov does coverage testing
- mock is used to mock the stripe APIs for our test
- fake-factory package used Faker underneath to generate fake data
- Flask-DebugToolbar is super useful - helps figure whats happening under the hood in your app.


Version 0.0.4 (2015-11-13)
^^^^^^^^^^^^^^^^^^^^^^^^^^

- Added CSRF protection.
- Added styles, fonts, images - beautified the frontend pages
- Added more frontend pages - faq, terms, privacy, learnmore etc.
- Implemented User models with complete user authentication - signup, login, password reset, change credentials etc
- Error templates - 404, 500 & 502 are served from build/public directory

Version 0.0.3 (2015-11-12)
^^^^^^^^^^^^^^^^^^^^^^^^^^

- Used Flask-Webpack to bundle all assets together.
- Very basic Jinja2 Templating done - used template inheritance & macros.
- Wrote the CHANGES.rst file

Version 0.0.2 (2015-11-10)
^^^^^^^^^^^^^^^^^^^^^^^^^^

- App now supports PostgreSQL database. The DB runs under a docker container
- gunicorn used to run the app & honcho Procfile used for automation.

Version 0.0.1 (2015-11-08)
^^^^^^^^^^^^^^^^^^^^^^^^^^

- Bootstrapped a simple Flask App with CLI


config/settings.py
# -*- coding: utf-8 -*-

from os import path
from datetime import timedelta

from celery.schedules import crontab


# This value is used for the following properties,
# it really should be your module's name.
#   Database name
#   Cache redis prefix
APP_NAME = 'lyfeshoppe'
APP_ROOT = path.join(path.dirname(path.abspath(__file__)), '..')

# App settings, most settings you see here will change in production.
SECRET_KEY = 'pickabettersecret'
DEBUG = True
TESTING = False
LOG_LEVEL = 'DEBUG'

# You will need to disable this to get Stripe's webhooks to work because you'll
# likely end up using tunneling tooling such as ngrok so the endpoints are
# reachable outside of your private network.
#
# The problem with this is, Flask won't allow any connections to the ngrok
# url with the SERVER_NAME set to localhost:8000. However if you comment out
# the SERVER_NAME below then webbooks will work but now url_for will not work
# inside of email templates.
#
# A better solution will turn up in the future.
SERVER_NAME = 'localhost:8000'

# Public build path. Files in this path will be accessible to the internet.
PUBLIC_BUILD_PATH = path.join(APP_ROOT, 'build', 'public')

# Flask-Webpack (assets) settings.
WEBPACK_MANIFEST_PATH = path.join(APP_ROOT, 'build', 'manifest.json')

# Babel i18n translations.
ACCEPT_LANGUAGES = ['en', 'es']
LANGUAGES = {
    'en': 'English',
    'es': u'Español'
}
BABEL_DEFAULT_LOCALE = 'en'

# Seed settings.
SEED_ADMIN_EMAIL = 'dev@localhost.com'

# Database settings,
# The username and password must match what's in docker-compose.yml for dev.
db_uri = 'postgresql://lyfeshoppe:bestpassword@localhost:5432/{0}'
SQLALCHEMY_DATABASE_URI = db_uri.format(APP_NAME)
SQLALCHEMY_POOL_SIZE = 5

# Cache settings.
CACHE_TYPE = 'redis'
CACHE_REDIS_URL = 'redis://localhost:6379/0'
CACHE_KEY_PREFIX = APP_NAME

# Celery settings.
CELERY_BROKER_URL = 'redis://localhost:6379/0'
CELERY_RESULT_BACKEND = 'redis://localhost:6379/0'
CELERY_REDIS_MAX_CONNECTIONS = 5
CELERY_ACCEPT_CONTENT = ['json']
CELERY_TASK_SERIALIZER = 'json'
CELERY_RESULT_SERIALIZER = 'json'

# Celery recurring scheduled tasks.
CELERYBEAT_SCHEDULE = {
    'mark-soon-to-expire-credit-cards': {
        'task': 'lyfeshoppe.blueprints.billing.tasks.mark_old_credit_cards',
        'schedule': crontab(hour=12, minute=1)
    },
    'mark-invalid-coupons': {
        'task': 'lyfeshoppe.blueprints.billing.tasks.expire_old_coupons',
        'schedule': crontab(hour=12, minute=2)
    },
}

# Login settings.
REMEMBER_COOKIE_DURATION = timedelta(days=90)

# Mail settings.
MAIL_DEFAULT_SENDER = 'support@lyfeshoppe.com'
MAIL_SERVER = 'smtp.gmail.com'
MAIL_PORT = 587
MAIL_USE_TLS = True
MAIL_USE_SSL = False

# External end points.
ENDPOINT_CADVISOR = 'http://localhost:8080/containers/'
ENDPOINT_FLOWER = 'http://localhost:8081'

# Stripe settings.
#
# API keys, NOTE: you should NOT supply them in this config, put them
# in an instance config file, such as: instance/settings.py
#
# They are only listed below to act as documentation.
STRIPE_SECRET_KEY = None
STRIPE_PUBLISHABLE_KEY = None

# The Stripe API version to use. More information can be found on their docs:
#  https://stripe.com/docs/api/python#versioning
STRIPE_API_VERSION = '2015-01-26'

# Documentation for each plan field below can be found on Stripe's API docs:
#  https://stripe.com/docs/api#create_plan
#
# After supplying both API keys and plans, you must sync the plans by running:
#   run stripe sync_plans
#
# If you are using TEST keys then the plans will be set to livemode: False
# If you are using REAL keys then the plans be set to livemode: True
#
# What the above means is, when you ship your app in production you must sync
# your plans at least once to activate them on your real keys.
STRIPE_PLANS = {
    '0': {
        'id': 'bronze',
        'name': 'Bronze',
        'amount': 100,
        'currency': 'usd',
        'interval': 'month',
        'interval_count': 1,
        'trial_period_days': 14,
        'statement_descriptor': 'BRONZE MONTHLY'
    },
    '1': {
        'id': 'gold',
        'name': 'Gold',
        'amount': 500,
        'currency': 'usd',
        'interval': 'month',
        'interval_count': 1,
        'trial_period_days': 14,
        'statement_descriptor': 'GOLD MONTHLY',
        'metadata': {
            'recommended': True
        }
    },
    '2': {
        'id': 'platinum',
        'name': 'Platinum',
        'amount': 1000,
        'currency': 'usd',
        'interval': 'month',
        'interval_count': 1,
        'trial_period_days': 14,
        'statement_descriptor': 'PLATINUM MONTHLY'
    }
}

# Twitter settings.
#
# API keys, NOTE: you should NOT supply them in this config, put them
# in an instance config file, such as: instance/settings.py
#
# They are only listed below to act as documentation.
TWITTER_CONSUMER_KEY = None
TWITTER_CONSUMER_SECRET = None
TWITTER_ACCESS_TOKEN = None
TWITTER_ACCESS_SECRET = None

# Broadcast (websocket server) settings.
#
# NOTE: you should NOT supply the PUSH_TOKEN/BROADCAST_INTERNAL_URL here,
# put them in an instance config file, such as: instance/settings.py
BROADCAST_PUBLIC_URL = 'http://localhost:4242/stream'
BROADCAST_INTERNAL_URL = None
BROADCAST_PUSH_TOKEN = None
